peps/pep-0788.rst

```diff
@@ -0,0 +1,686 @@
+    interpreter (created automatically by ``Py_Initialize()``). Python
+    supports the creation of additional interpreters (using
+    ``Py_NewInterpreter()``), but mixing multiple interpreters and the
+    ``PyGILState_*`` API is unsupported.
+
+More technically, this is because ``PyGILState_Ensure`` doesn't have any way
+to know which interpreter created the thread, and as such, it has to assume
+that it was the main interpreter. There isn't any way to detect this at
+runtime, so spurious races are bound to come up in threads created by
+subinterpreters, because synchronization for the wrong interpreter will be
+used on objects shared between the threads.
+
+
+Interpreters can concurrently shut down
+***************************************
+
+The other way of creating a native thread that can invoke Python,
+:c:func:`PyThreadState_New` / :c:func:`PyThreadState_Swap`, is a lot better
+for supporting subinterpreters (because :c:func:`PyThreadState_New` takes an
+explicit interpreter, rather than assuming that the main interpreter was intended),
+but is still limited by the current API.
+
+In particular, subinterpreters typically have a much shorter lifetime than the
+main interpreter, and as such, there's not necessarily a guarantee that a
+:c:type:`PyInterpreterState` (acquired by :c:func:`PyInterpreterState_Get`)
+passed to a fresh thread will still be alive. Similarly, a
+:c:type:`PyInterpreterState` pointer could have been replaced with a *new*
+interpreter, causing all sorts of unknown issues. They are also subject to
+all the finalization related hanging mentioned previously.
+
+Rationale
+=========
+
+This PEP includes several new APIs that intend to fix all of the issues stated
+above.
+
+Replacing the old APIs
+----------------------
+
+As made clear in Motivation_, ``PyGILState`` is already pretty buggy, and
+even if it was magically fixed, the current behavior of hanging the thread is
+beyond repair. In turn, this PEP intends to completely deprecate the existing
+``PyGILState`` APIs and provide better alternatives. However, even if this PEP
+is rejected, all of the APIs can be replaced with more correct ``PyThreadState``
+functions in the current C API:
+
+- :c:func:`PyGILState_Ensure`: :c:func:`PyThreadState_Swap` / :c:func:`PyThreadState_New`
+- :c:func:`PyGILState_Release`: :c:func:`PyThreadState_Clear` / :c:func:`PyThreadState_Delete`
+- :c:func:`PyGILState_GetThisThreadState`: :c:func:`PyThreadState_Get`
+- :c:func:`PyGILState_Check`: ``PyThreadState_GetUnchecked() != NULL``
+
+A light layer of magic
+----------------------
+
+The APIs proposed by this PEP intentionally have a layer of abstraction that is
+hidden from the user and offloads complexity onto CPython. This is done
+primarily to help ease the transition from ``PyGILState`` for existing
+codebases, and for ease-of-use to those who provide wrappers the C API, such
+as Cython or PyO3. See also :ref:`pep-788-activate-deactivate-instead`.
```

Comment:
This section says there is some mysterious abstraction, but it isn't clear what that is. Is it relevant to the reader to know this? If so, could you rephrase or add more context please?